File: .github/opencv5/smoke/cpp/main.cpp
#include <opencv2/calib.hpp>
#include <opencv2/core.hpp>
#include <opencv2/geometry.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/objdetect.hpp>

#include <cmath>
#include <iostream>
#include <string>
#include <vector>

int main() {
  const std::string version = cv::getVersionString();
  if (version != "5.0.0") {
    std::cerr << "unexpected OpenCV version: " << version << '\n';
    return 1;
  }

  const cv::Mat gray(64, 64, CV_8UC1, cv::Scalar::all(0));
  std::vector<unsigned char> png;
  if (!cv::imencode(".png", gray, png) || png.empty()) {
    std::cerr << "PNG codec smoke failed\n";
    return 2;
  }

  std::vector<cv::Point2f> corners;
  const bool found = cv::findChessboardCorners(
      gray, cv::Size(3, 3), corners,
      cv::CALIB_CB_ADAPTIVE_THRESH | cv::CALIB_CB_NORMALIZE_IMAGE);
  if (found) {
    std::cerr << "blank image unexpectedly contains a chessboard\n";
    return 3;
  }

  const std::vector<cv::Point3f> object_points{{0.0F, 0.0F, 1.0F}};
  const cv::Mat rotation = cv::Mat::zeros(3, 1, CV_64F);
  const cv::Mat translation = cv::Mat::zeros(3, 1, CV_64F);
  const cv::Mat camera_matrix = cv::Mat::eye(3, 3, CV_64F);
  const cv::Mat distortion = cv::Mat::zeros(1, 5, CV_64F);
  std::vector<cv::Point2f> projected;
  cv::projectPoints(object_points, rotation, translation, camera_matrix,
                    distortion, projected);
  if (projected.size() != 1 || !std::isfinite(projected[0].x) ||
      !std::isfinite(projected[0].y)) {
    std::cerr << "point projection smoke failed\n";
    return 4;
  }

  bool rejected_empty_calibration = false;
  try {
    const std::vector<std::vector<cv::Point3f>> empty_object_points;
    const std::vector<std::vector<cv::Point2f>> empty_image_points;
    cv::Mat mutable_camera_matrix = cv::Mat::eye(3, 3, CV_64F);
    cv::Mat mutable_distortion = cv::Mat::zeros(1, 5, CV_64F);
    std::vector<cv::Mat> rotations;
    std::vector<cv::Mat> translations;
    static_cast<void>(cv::calibrateCamera(
        empty_object_points, empty_image_points, cv::Size(64, 64),
        mutable_camera_matrix, mutable_distortion, rotations, translations));
  } catch (const cv::Exception&) {
    rejected_empty_calibration = true;
  }
  if (!rejected_empty_calibration) {
    std::cerr << "empty calibration input was not rejected\n";
    return 5;
  }

  std::cout << "OpenCV " << version << " relocated CMake smoke passed\n";
  return 0;
}

File: .github/opencv5/smoke/rust/src/main.rs
use opencv::calib;
use opencv::core::{self, Mat, Point2f, Point3f, Scalar, Size, TermCriteria, Vector};
use opencv::geometry;
use opencv::imgcodecs;
use opencv::objdetect;

fn main() -> opencv::Result<()> {
    let version = core::get_version_string()?;
    assert_eq!(version, "5.0.0", "unexpected OpenCV runtime version");

    let gray = Mat::new_rows_cols_with_default(64, 64, core::CV_8UC1, Scalar::all(0.0))?;
    let mut png = Vector::<u8>::new();
    assert!(
        imgcodecs::imencode_def(".png", &gray, &mut png)?,
        "PNG codec smoke failed"
    );
    assert!(!png.is_empty(), "PNG encoder returned an empty payload");

    let mut corners = Vector::<Point2f>::new();
    let found = objdetect::find_chessboard_corners(
        &gray,
        Size::new(3, 3),
        &mut corners,
        objdetect::CALIB_CB_ADAPTIVE_THRESH | objdetect::CALIB_CB_NORMALIZE_IMAGE,
    )?;
    assert!(!found, "blank image unexpectedly contains a chessboard");

    let object_points = Vector::<Point3f>::from_iter([Point3f::new(0.0, 0.0, 1.0)]);
    let rotation = Mat::from_slice_2d(&[&[0.0_f64], &[0.0], &[0.0]])?;
    let translation = Mat::from_slice_2d(&[&[0.0_f64], &[0.0], &[0.0]])?;
    let camera_matrix =
        Mat::from_slice_2d(&[&[1.0_f64, 0.0, 0.0], &[0.0, 1.0, 0.0], &[0.0, 0.0, 1.0]])?;
    let distortion = Mat::from_slice_2d(&[&[0.0_f64; 5]])?;
    let mut projected = Vector::<Point2f>::new();
    geometry::project_points_def(
        &object_points,
        &rotation,
        &translation,
        &camera_matrix,
        &distortion,
        &mut projected,
    )?;
    assert_eq!(projected.len(), 1, "point projection returned wrong size");
    let point = projected.get(0)?;
    assert!(
        point.x.is_finite() && point.y.is_finite(),
        "point projection returned non-finite coordinates"
    );

    let empty_object_points = Vector::<Vector<Point3f>>::new();
    let empty_image_points = Vector::<Vector<Point2f>>::new();
    let mut mutable_camera_matrix = camera_matrix.clone();
    let mut mutable_distortion = distortion.clone();
    let mut rotations = Vector::<Mat>::new();
    let mut translations = Vector::<Mat>::new();
    let criteria = TermCriteria::new(core::TermCriteria_COUNT | core::TermCriteria_EPS, 30, 1e-6)?;
    let calibration_result = calib::calibrate_camera(
        &empty_object_points,
        &empty_image_points,
        Size::new(64, 64),
        &mut mutable_camera_matrix,
        &mut mutable_distortion,
        &mut rotations,
        &mut translations,
        0,
        criteria,
    );
    assert!(
        calibration_result.is_err(),
        "empty calibration input was not rejected"
    );

    println!("OpenCV {version} relocated opencv-rust smoke passed");
    Ok(())
}
